Unit tests should not require credentials from environment

```bash
$ git remote -v
origin	git@github.com:googleapis/python-spanner (fetch)
origin	git@github.com:googleapis/python-spanner (push)

$ git status
On branch master
Your branch is up to date with 'origin/master'.

nothing to commit, working tree clean

$ env | grep GOOGLE || echo "No credentials in environment"
No credentials in environment

$ .nox/unit-3-6/bin/py.test -v -x tests/unit/spanner_dbapi/
...
=================================== FAILURES ===================================
____________________ Test_connect.test_connect_database_id _____________________

self = <tests.unit.spanner_dbapi.test_connect.Test_connect testMethod=test_connect_database_id>

    def test_connect_database_id(self):
        from google.cloud.spanner_dbapi import connect
        from google.cloud.spanner_dbapi import Connection
    
        DATABASE = "test-database"
    
        with mock.patch(
            "google.cloud.spanner_v1.instance.Instance.database"
        ) as database_mock:
            with mock.patch(
                "google.cloud.spanner_v1.instance.Instance.exists", return_value=True,
            ):
>               connection = connect("test-instance", DATABASE)

tests/unit/spanner_dbapi/test_connect.py:110: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
google/cloud/spanner_dbapi/connection.py:399: in connect
    project=project, credentials=credentials, client_info=client_info
google/cloud/spanner_v1/client.py:162: in __init__
    _http=None,
.nox/unit-3-6/lib/python3.6/site-packages/google/cloud/client.py:249: in __init__
    _ClientProjectMixin.__init__(self, project=project)
.nox/unit-3-6/lib/python3.6/site-packages/google/cloud/client.py:201: in __init__
    project = self._determine_default(project)
.nox/unit-3-6/lib/python3.6/site-packages/google/cloud/client.py:216: in _determine_default
    return _determine_default_project(project)
.nox/unit-3-6/lib/python3.6/site-packages/google/cloud/_helpers.py:186: in _determine_default_project
    _, project = google.auth.default()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

scopes = None, request = None, quota_project_id = None, default_scopes = None

    def default(scopes=None, request=None, quota_project_id=None, default_scopes=None):
        """Gets the default credentials for the current environment.
    
        `Application Default Credentials`_ provides an easy way to obtain
        credentials to call Google APIs for server-to-server or local applications.
        This function acquires credentials from the environment in the following
        order:
    
        1. If the environment variable ``GOOGLE_APPLICATION_CREDENTIALS`` is set
           to the path of a valid service account JSON private key file, then it is
           loaded and returned. The project ID returned is the project ID defined
           in the service account file if available (some older files do not
           contain project ID information).
    
           If the environment variable is set to the path of a valid external
           account JSON configuration file (workload identity federation), then the
           configuration file is used to determine and retrieve the external
           credentials from the current environment (AWS, Azure, etc).
           These will then be exchanged for Google access tokens via the Google STS
           endpoint.
           The project ID returned in this case is the one corresponding to the
           underlying workload identity pool resource if determinable.
        2. If the `Google Cloud SDK`_ is installed and has application default
           credentials set they are loaded and returned.
    
           To enable application default credentials with the Cloud SDK run::
    
                gcloud auth application-default login
    
           If the Cloud SDK has an active project, the project ID is returned. The
           active project can be set using::
    
                gcloud config set project
    
        3. If the application is running in the `App Engine standard environment`_
           (first generation) then the credentials and project ID from the
           `App Identity Service`_ are used.
        4. If the application is running in `Compute Engine`_ or `Cloud Run`_ or
           the `App Engine flexible environment`_ or the `App Engine standard
           environment`_ (second generation) then the credentials and project ID
           are obtained from the `Metadata Service`_.
        5. If no credentials are found,
           :class:`~google.auth.exceptions.DefaultCredentialsError` will be raised.
    
        .. _Application Default Credentials: https://developers.google.com\
                /identity/protocols/application-default-credentials
        .. _Google Cloud SDK: https://cloud.google.com/sdk
        .. _App Engine standard environment: https://cloud.google.com/appengine
        .. _App Identity Service: https://cloud.google.com/appengine/docs/python\
                /appidentity/
        .. _Compute Engine: https://cloud.google.com/compute
        .. _App Engine flexible environment: https://cloud.google.com\
                /appengine/flexible
        .. _Metadata Service: https://cloud.google.com/compute/docs\
                /storing-retrieving-metadata
        .. _Cloud Run: https://cloud.google.com/run
    
        Example::
    
            import google.auth
    
            credentials, project_id = google.auth.default()
    
        Args:
            scopes (Sequence[str]): The list of scopes for the credentials. If
                specified, the credentials will automatically be scoped if
                necessary.
            request (Optional[google.auth.transport.Request]): An object used to make
                HTTP requests. This is used to either detect whether the application
                is running on Compute Engine or to determine the associated project
                ID for a workload identity pool resource (external account
                credentials). If not specified, then it will either use the standard
                library http client to make requests for Compute Engine credentials
                or a google.auth.transport.requests.Request client for external
                account credentials.
            quota_project_id (Optional[str]): The project ID used for
                quota and billing.
            default_scopes (Optional[Sequence[str]]): Default scopes passed by a
                Google client library. Use 'scopes' for user-defined scopes.
        Returns:
            Tuple[~google.auth.credentials.Credentials, Optional[str]]:
                the current environment's credentials and project ID. Project ID
                may be None, which indicates that the Project ID could not be
                ascertained from the environment.
    
        Raises:
            ~google.auth.exceptions.DefaultCredentialsError:
                If no credentials were found, or if the credentials found were
                invalid.
        """
        from google.auth.credentials import with_scopes_if_required
    
        explicit_project_id = os.environ.get(
            environment_vars.PROJECT, os.environ.get(environment_vars.LEGACY_PROJECT)
        )
    
        checkers = (
            # Avoid passing scopes here to prevent passing scopes to user credentials.
            # with_scopes_if_required() below will ensure scopes/default scopes are
            # safely set on the returned credentials since requires_scopes will
            # guard against setting scopes on user credentials.
            _get_explicit_environ_credentials,
            _get_gcloud_sdk_credentials,
            _get_gae_credentials,
            lambda: _get_gce_credentials(request),
        )
    
        for checker in checkers:
            credentials, project_id = checker()
            if credentials is not None:
                credentials = with_scopes_if_required(
                    credentials, scopes, default_scopes=default_scopes
                )
    
                # For external account credentials, scopes are required to determine
                # the project ID. Try to get the project ID again if not yet
                # determined.
                if not project_id and callable(
                    getattr(credentials, "get_project_id", None)
                ):
                    if request is None:
                        request = google.auth.transport.requests.Request()
                    project_id = credentials.get_project_id(request=request)
    
                if quota_project_id:
                    credentials = credentials.with_quota_project(quota_project_id)
    
                effective_project_id = explicit_project_id or project_id
                if not effective_project_id:
                    _LOGGER.warning(
                        "No project ID could be determined. Consider running "
                        "`gcloud config set project` or setting the %s "
                        "environment variable",
                        environment_vars.PROJECT,
                    )
                return credentials, effective_project_id
    
>       raise exceptions.DefaultCredentialsError(_HELP_MESSAGE)
E       google.auth.exceptions.DefaultCredentialsError: Could not automatically determine credentials. Please set GOOGLE_APPLICATION_CREDENTIALS or explicitly create credentials and re-run the application. For more information, please see https://cloud.google.com/docs/authentication/getting-started

.nox/unit-3-6/lib/python3.6/site-packages/google/auth/_default.py:488: DefaultCredentialsError
------------------------------ Captured log call -------------------------------
WARNING  google.auth.compute_engine._metadata:_metadata.py:104 Compute Engine Metadata server unavailable onattempt 1 of 3. Reason: timed out
WARNING  google.auth.compute_engine._metadata:_metadata.py:104 Compute Engine Metadata server unavailable onattempt 2 of 3. Reason: [Errno 113] No route to host
WARNING  google.auth.compute_engine._metadata:_metadata.py:104 Compute Engine Metadata server unavailable onattempt 3 of 3. Reason: timed out
WARNING  google.auth._default:_default.py:287 Authentication failed using Compute Engine authentication due to unavailable metadata server.
=========================== short test summary info ============================
FAILED tests/unit/spanner_dbapi/test_connect.py::Test_connect::test_connect_database_id
!!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!!
========================= 1 failed, 11 passed in 6.81s =========================
```

--- a/tests/unit/spanner_dbapi/test_connect.py
+++ b/tests/unit/spanner_dbapi/test_connect.py
@@ -20,6 +20,12 @@ from unittest import mock
 import google.auth.credentials
 
 
+INSTANCE = "test-instance"
+DATABASE = "test-database"
+PROJECT = "test-project"
+USER_AGENT = "user-agent"
+
+
 def _make_credentials():
     class _CredentialsWithScopes(
         google.auth.credentials.Credentials, google.auth.credentials.Scoped
@@ -29,138 +35,105 @@ def _make_credentials():
     return mock.Mock(spec=_CredentialsWithScopes)
 
 
+@mock.patch("google.cloud.spanner_v1.Client")
 class Test_connect(unittest.TestCase):
-    def test_connect(self):
+    def test_w_implicit(self, mock_client):
         from google.cloud.spanner_dbapi import connect
         from google.cloud.spanner_dbapi import Connection
 
-        PROJECT = "test-project"
-        USER_AGENT = "user-agent"
-        CREDENTIALS = _make_credentials()
+        client = mock_client.return_value
+        instance = client.instance.return_value
+        database = instance.database.return_value
 
-        with mock.patch("google.cloud.spanner_v1.Client") as client_mock:
-            connection = connect(
-                "test-instance",
-                "test-database",
-                PROJECT,
-                CREDENTIALS,
-                user_agent=USER_AGENT,
-            )
-
-            self.assertIsInstance(connection, Connection)
-
-            client_mock.assert_called_once_with(
-                project=PROJECT, credentials=CREDENTIALS, client_info=mock.ANY
-            )
-
-    def test_instance_not_found(self):
-        from google.cloud.spanner_dbapi import connect
-
-        with mock.patch(
-            "google.cloud.spanner_v1.instance.Instance.exists", return_value=False,
-        ) as exists_mock:
-
-            with self.assertRaises(ValueError):
-                connect("test-instance", "test-database")
-
-            exists_mock.assert_called_once_with()
-
-    def test_database_not_found(self):
-        from google.cloud.spanner_dbapi import connect
-
-        with mock.patch(
-            "google.cloud.spanner_v1.instance.Instance.exists", return_value=True,
-        ):
-            with mock.patch(
-                "google.cloud.spanner_v1.database.Database.exists", return_value=False,
-            ) as exists_mock:
-
-                with self.assertRaises(ValueError):
-                    connect("test-instance", "test-database")
-
-                exists_mock.assert_called_once_with()
-
-    def test_connect_instance_id(self):
-        from google.cloud.spanner_dbapi import connect
-        from google.cloud.spanner_dbapi import Connection
-
-        INSTANCE = "test-instance"
-
-        with mock.patch(
-            "google.cloud.spanner_v1.client.Client.instance"
-        ) as instance_mock:
-            connection = connect(INSTANCE, "test-database")
-
-            instance_mock.assert_called_once_with(INSTANCE)
+        connection = connect(INSTANCE, DATABASE)
 
         self.assertIsInstance(connection, Connection)
 
-    def test_connect_database_id(self):
+        self.assertIs(connection.instance, instance)
+        client.instance.assert_called_once_with(INSTANCE)
+
+        self.assertIs(connection.database, database)
+        instance.database.assert_called_once_with(DATABASE, pool=None)
+        # Datbase constructs its own pool
+        self.assertIsNotNone(connection.database._pool)
+
+    def test_w_explicit(self, mock_client):
+        from google.cloud.spanner_v1.pool import AbstractSessionPool
         from google.cloud.spanner_dbapi import connect
         from google.cloud.spanner_dbapi import Connection
+        from google.cloud.spanner_dbapi.version import PY_VERSION
 
-        DATABASE = "test-database"
+        credentials = _make_credentials()
+        pool = mock.create_autospec(AbstractSessionPool)
+        client = mock_client.return_value
+        instance = client.instance.return_value
+        database = instance.database.return_value
 
-        with mock.patch(
-            "google.cloud.spanner_v1.instance.Instance.database"
-        ) as database_mock:
-            with mock.patch(
-                "google.cloud.spanner_v1.instance.Instance.exists", return_value=True,
-            ):
-                connection = connect("test-instance", DATABASE)
-
-                database_mock.assert_called_once_with(DATABASE, pool=mock.ANY)
+        connection = connect(
+            INSTANCE, DATABASE, PROJECT, credentials, pool=pool, user_agent=USER_AGENT,
+        )
 
         self.assertIsInstance(connection, Connection)
 
-    def test_default_sessions_pool(self):
+        mock_client.assert_called_once_with(
+            project=PROJECT, credentials=credentials, client_info=mock.ANY
+        )
+        client_info = mock_client.call_args_list[0][1]["client_info"]
+        self.assertEqual(client_info.user_agent, USER_AGENT)
+        self.assertEqual(client_info.python_version, PY_VERSION)
+
+        self.assertIs(connection.instance, instance)
+        client.instance.assert_called_once_with(INSTANCE)
+
+        self.assertIs(connection.database, database)
+        instance.database.assert_called_once_with(DATABASE, pool=pool)
+
+    def test_w_instance_not_found(self, mock_client):
         from google.cloud.spanner_dbapi import connect
 
-        with mock.patch("google.cloud.spanner_v1.instance.Instance.database"):
-            with mock.patch(
-                "google.cloud.spanner_v1.instance.Instance.exists", return_value=True,
-            ):
-                connection = connect("test-instance", "test-database")
+        client = mock_client.return_value
+        instance = client.instance.return_value
+        instance.exists.return_value = False
 
-                self.assertIsNotNone(connection.database._pool)
+        with self.assertRaises(ValueError):
+            connect(INSTANCE, DATABASE)
 
-    def test_sessions_pool(self):
+        instance.exists.assert_called_once_with()
+
+    def test_w_database_not_found(self, mock_client):
         from google.cloud.spanner_dbapi import connect
-        from google.cloud.spanner_v1.pool import FixedSizePool
 
-        database_id = "test-database"
-        pool = FixedSizePool()
+        client = mock_client.return_value
+        instance = client.instance.return_value
+        database = instance.database.return_value
+        database.exists.return_value = False
 
-        with mock.patch(
-            "google.cloud.spanner_v1.instance.Instance.database"
-        ) as database_mock:
-            with mock.patch(
-                "google.cloud.spanner_v1.instance.Instance.exists", return_value=True,
-            ):
-                connect("test-instance", database_id, pool=pool)
-                database_mock.assert_called_once_with(database_id, pool=pool)
+        with self.assertRaises(ValueError):
+            connect(INSTANCE, DATABASE)
 
-    def test_connect_w_credential_file_path(self):
+        database.exists.assert_called_once_with()
+
+    def test_w_credential_file_path(self, mock_client):
         from google.cloud.spanner_dbapi import connect
         from google.cloud.spanner_dbapi import Connection
+        from google.cloud.spanner_dbapi.version import PY_VERSION
 
-        PROJECT = "test-project"
-        USER_AGENT = "user-agent"
-        credentials = "dummy/file/path.json"
+        credentials_path = "dummy/file/path.json"
 
-        with mock.patch(
-            "google.cloud.spanner_v1.Client.from_service_account_json"
-        ) as client_mock:
-            connection = connect(
-                "test-instance",
-                "test-database",
-                PROJECT,
-                credentials=credentials,
-                user_agent=USER_AGENT,
-            )
+        connection = connect(
+            INSTANCE,
+            DATABASE,
+            PROJECT,
+            credentials=credentials_path,
+            user_agent=USER_AGENT,
+        )
 
-            self.assertIsInstance(connection, Connection)
+        self.assertIsInstance(connection, Connection)
 
-            client_mock.assert_called_once_with(
-                credentials, project=PROJECT, client_info=mock.ANY
-            )
+        factory = mock_client.from_service_account_json
+        factory.assert_called_once_with(
+            credentials_path, project=PROJECT, client_info=mock.ANY,
+        )
+        client_info = factory.call_args_list[0][1]["client_info"]
+        self.assertEqual(client_info.user_agent, USER_AGENT)
+        self.assertEqual(client_info.python_version, PY_VERSION)

--- a/tests/unit/spanner_dbapi/test_connection.py
+++ b/tests/unit/spanner_dbapi/test_connection.py
@@ -18,6 +18,11 @@ import mock
 import unittest
 import warnings
 
+PROJECT = "test-project"
+INSTANCE = "test-instance"
+DATABASE = "test-database"
+USER_AGENT = "user-agent"
+
 
 def _make_credentials():
     from google.auth import credentials
@@ -29,78 +34,62 @@ def _make_credentials():
 
 
 class TestConnection(unittest.TestCase):
-
-    PROJECT = "test-project"
-    INSTANCE = "test-instance"
-    DATABASE = "test-database"
-    USER_AGENT = "user-agent"
-    CREDENTIALS = _make_credentials()
-
     def _get_client_info(self):
         from google.api_core.gapic_v1.client_info import ClientInfo
 
-        return ClientInfo(user_agent=self.USER_AGENT)
+        return ClientInfo(user_agent=USER_AGENT)
 
     def _make_connection(self):
         from google.cloud.spanner_dbapi import Connection
         from google.cloud.spanner_v1.instance import Instance
 
         # We don't need a real Client object to test the constructor
-        instance = Instance(self.INSTANCE, client=None)
-        database = instance.database(self.DATABASE)
+        instance = Instance(INSTANCE, client=None)
+        database = instance.database(DATABASE)
         return Connection(instance, database)
 
-    def test_autocommit_setter_transaction_not_started(self):
+    @mock.patch("google.cloud.spanner_dbapi.connection.Connection.commit")
+    def test_autocommit_setter_transaction_not_started(self, mock_commit):
         connection = self._make_connection()
 
-        with mock.patch(
-            "google.cloud.spanner_dbapi.connection.Connection.commit"
-        ) as mock_commit:
-            connection.autocommit = True
-            mock_commit.assert_not_called()
-            self.assertTrue(connection._autocommit)
+        connection.autocommit = True
 
-        with mock.patch(
-            "google.cloud.spanner_dbapi.connection.Connection.commit"
-        ) as mock_commit:
-            connection.autocommit = False
-            mock_commit.assert_not_called()
-            self.assertFalse(connection._autocommit)
+        mock_commit.assert_not_called()
+        self.assertTrue(connection._autocommit)
 
-    def test_autocommit_setter_transaction_started(self):
+        connection.autocommit = False
+        mock_commit.assert_not_called()
+        self.assertFalse(connection._autocommit)
+
+    @mock.patch("google.cloud.spanner_dbapi.connection.Connection.commit")
+    def test_autocommit_setter_transaction_started(self, mock_commit):
+        connection = self._make_connection()
+        connection._transaction = mock.Mock(committed=False, rolled_back=False)
+
+        connection.autocommit = True
+
+        mock_commit.assert_called_once()
+        self.assertTrue(connection._autocommit)
+
+    @mock.patch("google.cloud.spanner_dbapi.connection.Connection.commit")
+    def test_autocommit_setter_transaction_started_commited_rolled_back(
+        self, mock_commit
+    ):
         connection = self._make_connection()
 
-        with mock.patch(
-            "google.cloud.spanner_dbapi.connection.Connection.commit"
-        ) as mock_commit:
-            connection._transaction = mock.Mock(committed=False, rolled_back=False)
+        connection._transaction = mock.Mock(committed=True, rolled_back=False)
 
-            connection.autocommit = True
-            mock_commit.assert_called_once()
-            self.assertTrue(connection._autocommit)
-
-    def test_autocommit_setter_transaction_started_commited_rolled_back(self):
-        connection = self._make_connection()
-
-        with mock.patch(
-            "google.cloud.spanner_dbapi.connection.Connection.commit"
-        ) as mock_commit:
-            connection._transaction = mock.Mock(committed=True, rolled_back=False)
-
-            connection.autocommit = True
-            mock_commit.assert_not_called()
-            self.assertTrue(connection._autocommit)
+        connection.autocommit = True
+        mock_commit.assert_not_called()
+        self.assertTrue(connection._autocommit)
 
         connection.autocommit = False
 
-        with mock.patch(
-            "google.cloud.spanner_dbapi.connection.Connection.commit"
-        ) as mock_commit:
-            connection._transaction = mock.Mock(committed=False, rolled_back=True)
+        connection._transaction = mock.Mock(committed=False, rolled_back=True)
 
-            connection.autocommit = True
-            mock_commit.assert_not_called()
-            self.assertTrue(connection._autocommit)
+        connection.autocommit = True
+        mock_commit.assert_not_called()
+        self.assertTrue(connection._autocommit)
 
     def test_property_database(self):
         from google.cloud.spanner_v1.database import Database
@@ -116,76 +105,92 @@ class TestConnection(unittest.TestCase):
         self.assertIsInstance(connection.instance, Instance)
         self.assertEqual(connection.instance, connection._instance)
 
-    def test__session_checkout(self):
+    @staticmethod
+    def _make_pool():
+        from google.cloud.spanner_v1.pool import AbstractSessionPool
+
+        return mock.create_autospec(AbstractSessionPool)
+
+    @mock.patch("google.cloud.spanner_v1.database.Database")
+    def test__session_checkout(self, mock_database):
         from google.cloud.spanner_dbapi import Connection
 
-        with mock.patch("google.cloud.spanner_v1.database.Database") as mock_database:
-            mock_database._pool = mock.MagicMock()
-            mock_database._pool.get = mock.MagicMock(return_value="db_session_pool")
-            connection = Connection(self.INSTANCE, mock_database)
+        pool = self._make_pool()
+        mock_database._pool = pool
+        connection = Connection(INSTANCE, mock_database)
 
-            connection._session_checkout()
-            mock_database._pool.get.assert_called_once_with()
-            self.assertEqual(connection._session, "db_session_pool")
+        connection._session_checkout()
+        pool.get.assert_called_once_with()
+        self.assertEqual(connection._session, pool.get.return_value)
 
-            connection._session = "db_session"
-            connection._session_checkout()
-            self.assertEqual(connection._session, "db_session")
+        connection._session = "db_session"
+        connection._session_checkout()
+        self.assertEqual(connection._session, "db_session")
 
-    def test__release_session(self):
+    @mock.patch("google.cloud.spanner_v1.database.Database")
+    def test__release_session(self, mock_database):
         from google.cloud.spanner_dbapi import Connection
 
-        with mock.patch("google.cloud.spanner_v1.database.Database") as mock_database:
-            mock_database._pool = mock.MagicMock()
-            mock_database._pool.put = mock.MagicMock()
-            connection = Connection(self.INSTANCE, mock_database)
-            connection._session = "session"
+        pool = self._make_pool()
+        mock_database._pool = pool
+        connection = Connection(INSTANCE, mock_database)
+        connection._session = "session"
 
-            connection._release_session()
-            mock_database._pool.put.assert_called_once_with("session")
-            self.assertIsNone(connection._session)
+        connection._release_session()
+        pool.put.assert_called_once_with("session")
+        self.assertIsNone(connection._session)
 
     def test_transaction_checkout(self):
         from google.cloud.spanner_dbapi import Connection
 
-        connection = Connection(self.INSTANCE, self.DATABASE)
-        connection._session_checkout = mock_checkout = mock.MagicMock(autospec=True)
+        connection = Connection(INSTANCE, DATABASE)
+        mock_checkout = mock.MagicMock(autospec=True)
+        connection._session_checkout = mock_checkout
+
         connection.transaction_checkout()
+
         mock_checkout.assert_called_once_with()
 
-        connection._transaction = mock_transaction = mock.MagicMock()
+        mock_transaction = mock.MagicMock()
         mock_transaction.committed = mock_transaction.rolled_back = False
+        connection._transaction = mock_transaction
+
         self.assertEqual(connection.transaction_checkout(), mock_transaction)
 
         connection._autocommit = True
         self.assertIsNone(connection.transaction_checkout())
 
-    def test_close(self):
-        from google.cloud.spanner_dbapi import connect, InterfaceError
+    @mock.patch("google.cloud.spanner_v1.Client")
+    def test_close(self, mock_client):
+        from google.cloud.spanner_dbapi import connect
+        from google.cloud.spanner_dbapi import InterfaceError
 
-        with mock.patch(
-            "google.cloud.spanner_v1.instance.Instance.exists", return_value=True
-        ):
-            with mock.patch(
-                "google.cloud.spanner_v1.database.Database.exists", return_value=True
-            ):
-                connection = connect("test-instance", "test-database")
+        connection = connect("test-instance", "test-database")
 
         self.assertFalse(connection.is_closed)
+
         connection.close()
+
         self.assertTrue(connection.is_closed)
 
         with self.assertRaises(InterfaceError):
             connection.cursor()
 
-        connection._transaction = mock_transaction = mock.MagicMock()
+        mock_transaction = mock.MagicMock()
         mock_transaction.committed = mock_transaction.rolled_back = False
-        mock_transaction.rollback = mock_rollback = mock.MagicMock()
+        connection._transaction = mock_transaction
+
+        mock_rollback = mock.MagicMock()
+        mock_transaction.rollback = mock_rollback
+
         connection.close()
+
         mock_rollback.assert_called_once_with()
+
         connection._transaction = mock.MagicMock()
         connection._own_pool = False
         connection.close()
+
         self.assertTrue(connection.is_closed)
 
     @mock.patch.object(warnings, "warn")
@@ -193,13 +198,14 @@ class TestConnection(unittest.TestCase):
         from google.cloud.spanner_dbapi import Connection
         from google.cloud.spanner_dbapi.connection import AUTOCOMMIT_MODE_WARNING
 
-        connection = Connection(self.INSTANCE, self.DATABASE)
+        connection = Connection(INSTANCE, DATABASE)
 
         with mock.patch(
             "google.cloud.spanner_dbapi.connection.Connection._release_session"
         ) as mock_release:
             connection.commit()
-            mock_release.assert_not_called()
+
+        mock_release.assert_not_called()
 
         connection._transaction = mock_transaction = mock.MagicMock(
             rolled_back=False, committed=False
@@ -210,8 +216,9 @@ class TestConnection(unittest.TestCase):
             "google.cloud.spanner_dbapi.connection.Connection._release_session"
         ) as mock_release:
             connection.commit()
-            mock_commit.assert_called_once_with()
-            mock_release.assert_called_once_with()
+
+        mock_commit.assert_called_once_with()
+        mock_release.assert_called_once_with()
 
         connection._autocommit = True
         connection.commit()
@@ -224,23 +231,27 @@ class TestConnection(unittest.TestCase):
         from google.cloud.spanner_dbapi import Connection
         from google.cloud.spanner_dbapi.connection import AUTOCOMMIT_MODE_WARNING
 
-        connection = Connection(self.INSTANCE, self.DATABASE)
+        connection = Connection(INSTANCE, DATABASE)
 
         with mock.patch(
             "google.cloud.spanner_dbapi.connection.Connection._release_session"
         ) as mock_release:
             connection.rollback()
-            mock_release.assert_not_called()
 
-        connection._transaction = mock_transaction = mock.MagicMock()
-        mock_transaction.rollback = mock_rollback = mock.MagicMock()
+        mock_release.assert_not_called()
+
+        mock_transaction = mock.MagicMock()
+        connection._transaction = mock_transaction
+        mock_rollback = mock.MagicMock()
+        mock_transaction.rollback = mock_rollback
 
         with mock.patch(
             "google.cloud.spanner_dbapi.connection.Connection._release_session"
         ) as mock_release:
             connection.rollback()
-            mock_rollback.assert_called_once_with()
-            mock_release.assert_called_once_with()
+
+        mock_rollback.assert_called_once_with()
+        mock_release.assert_called_once_with()
 
         connection._autocommit = True
         connection.rollback()
@@ -248,101 +259,34 @@ class TestConnection(unittest.TestCase):
             AUTOCOMMIT_MODE_WARNING, UserWarning, stacklevel=2
         )
 
-    def test_run_prior_DDL_statements(self):
+    @mock.patch("google.cloud.spanner_v1.database.Database", autospec=True)
+    def test_run_prior_DDL_statements(self, mock_database):
         from google.cloud.spanner_dbapi import Connection, InterfaceError
 
-        with mock.patch(
-            "google.cloud.spanner_v1.database.Database", autospec=True
-        ) as mock_database:
-            connection = Connection(self.INSTANCE, mock_database)
+        connection = Connection(INSTANCE, mock_database)
 
+        connection.run_prior_DDL_statements()
+        mock_database.update_ddl.assert_not_called()
+
+        ddl = ["ddl"]
+        connection._ddl_statements = ddl
+
+        connection.run_prior_DDL_statements()
+        mock_database.update_ddl.assert_called_once_with(ddl)
+
+        connection.is_closed = True
+
+        with self.assertRaises(InterfaceError):
             connection.run_prior_DDL_statements()
-            mock_database.update_ddl.assert_not_called()
 
-            ddl = ["ddl"]
-            connection._ddl_statements = ddl
-
-            connection.run_prior_DDL_statements()
-            mock_database.update_ddl.assert_called_once_with(ddl)
-
-            connection.is_closed = True
-
-            with self.assertRaises(InterfaceError):
-                connection.run_prior_DDL_statements()
-
-    def test_context(self):
+    def test_as_context_manager(self):
         connection = self._make_connection()
         with connection as conn:
             self.assertEqual(conn, connection)
 
         self.assertTrue(connection.is_closed)
 
-    def test_connect(self):
-        from google.cloud.spanner_dbapi import Connection, connect
-
-        with mock.patch("google.cloud.spanner_v1.Client"):
-            with mock.patch(
-                "google.api_core.gapic_v1.client_info.ClientInfo",
-                return_value=self._get_client_info(),
-            ):
-                connection = connect(
-                    self.INSTANCE,
-                    self.DATABASE,
-                    self.PROJECT,
-                    self.CREDENTIALS,
-                    self.USER_AGENT,
-                )
-                self.assertIsInstance(connection, Connection)
-
-    def test_connect_instance_not_found(self):
-        from google.cloud.spanner_dbapi import connect
-
-        with mock.patch(
-            "google.cloud.spanner_v1.instance.Instance.exists", return_value=False
-        ):
-            with self.assertRaises(ValueError):
-                connect("test-instance", "test-database")
-
-    def test_connect_database_not_found(self):
-        from google.cloud.spanner_dbapi import connect
-
-        with mock.patch(
-            "google.cloud.spanner_v1.database.Database.exists", return_value=False
-        ):
-            with mock.patch(
-                "google.cloud.spanner_v1.instance.Instance.exists", return_value=True
-            ):
-                with self.assertRaises(ValueError):
-                    connect("test-instance", "test-database")
-
-    def test_default_sessions_pool(self):
-        from google.cloud.spanner_dbapi import connect
-
-        with mock.patch("google.cloud.spanner_v1.instance.Instance.database"):
-            with mock.patch(
-                "google.cloud.spanner_v1.instance.Instance.exists", return_value=True
-            ):
-                connection = connect("test-instance", "test-database")
-
-                self.assertIsNotNone(connection.database._pool)
-
-    def test_sessions_pool(self):
-        from google.cloud.spanner_dbapi import connect
-        from google.cloud.spanner_v1.pool import FixedSizePool
-
-        database_id = "test-database"
-        pool = FixedSizePool()
-
-        with mock.patch(
-            "google.cloud.spanner_v1.instance.Instance.database"
-        ) as database_mock:
-            with mock.patch(
-                "google.cloud.spanner_v1.instance.Instance.exists", return_value=True
-            ):
-                connect("test-instance", database_id, pool=pool)
-                database_mock.assert_called_once_with(database_id, pool=pool)
-
-    def test_run_statement_remember_statements(self):
+    def test_run_statement_wo_retried(self):
         """Check that Connection remembers executed statements."""
         from google.cloud.spanner_dbapi.checksum import ResultsChecksum
         from google.cloud.spanner_dbapi.cursor import Statement
@@ -352,19 +296,16 @@ class TestConnection(unittest.TestCase):
         param_types = {"a1": str}
 
         connection = self._make_connection()
-
+        connection.transaction_checkout = mock.Mock()
         statement = Statement(sql, params, param_types, ResultsChecksum(), False)
-        with mock.patch(
-            "google.cloud.spanner_dbapi.connection.Connection.transaction_checkout"
-        ):
-            connection.run_statement(statement)
+        connection.run_statement(statement)
 
         self.assertEqual(connection._statements[0].sql, sql)
         self.assertEqual(connection._statements[0].params, params)
         self.assertEqual(connection._statements[0].param_types, param_types)
         self.assertIsInstance(connection._statements[0].checksum, ResultsChecksum)
 
-    def test_run_statement_dont_remember_retried_statements(self):
+    def test_run_statement_w_retried(self):
         """Check that Connection doesn't remember re-executed statements."""
         from google.cloud.spanner_dbapi.checksum import ResultsChecksum
         from google.cloud.spanner_dbapi.cursor import Statement
@@ -374,12 +315,9 @@ class TestConnection(unittest.TestCase):
         param_types = {"a1": str}
 
         connection = self._make_connection()
-
+        connection.transaction_checkout = mock.Mock()
         statement = Statement(sql, params, param_types, ResultsChecksum(), False)
-        with mock.patch(
-            "google.cloud.spanner_dbapi.connection.Connection.transaction_checkout"
-        ):
-            connection.run_statement(statement, retried=True)
+        connection.run_statement(statement, retried=True)
 
         self.assertEqual(len(connection._statements), 0)
 
@@ -393,12 +331,10 @@ class TestConnection(unittest.TestCase):
         param_types = None
 
         connection = self._make_connection()
-
+        connection.transaction_checkout = mock.Mock()
         statement = Statement(sql, params, param_types, ResultsChecksum(), True)
-        with mock.patch(
-            "google.cloud.spanner_dbapi.connection.Connection.transaction_checkout"
-        ):
-            connection.run_statement(statement, retried=True)
+
+        connection.run_statement(statement, retried=True)
 
         self.assertEqual(len(connection._statements), 0)
 
@@ -412,16 +348,15 @@ class TestConnection(unittest.TestCase):
         param_types = {"f1": str, "f2": str}
 
         connection = self._make_connection()
-
+        connection.transaction_checkout = mock.Mock()
         statement = Statement(sql, params, param_types, ResultsChecksum(), True)
-        with mock.patch(
-            "google.cloud.spanner_dbapi.connection.Connection.transaction_checkout"
-        ):
-            connection.run_statement(statement, retried=True)
+
+        connection.run_statement(statement, retried=True)
 
         self.assertEqual(len(connection._statements), 0)
 
-    def test_clear_statements_on_commit(self):
+    @mock.patch("google.cloud.spanner_v1.transaction.Transaction")
+    def test_commit_clears_statements(self, mock_transaction):
         """
         Check that all the saved statements are
         cleared, when the transaction is commited.
@@ -432,12 +367,12 @@ class TestConnection(unittest.TestCase):
 
         self.assertEqual(len(connection._statements), 2)
 
-        with mock.patch("google.cloud.spanner_v1.transaction.Transaction.commit"):
-            connection.commit()
+        connection.commit()
 
         self.assertEqual(len(connection._statements), 0)
 
-    def test_clear_statements_on_rollback(self):
+    @mock.patch("google.cloud.spanner_v1.transaction.Transaction")
+    def test_rollback_clears_statements(self, mock_transaction):
         """
         Check that all the saved statements are
         cleared, when the transaction is roll backed.
@@ -448,40 +383,36 @@ class TestConnection(unittest.TestCase):
 
         self.assertEqual(len(connection._statements), 2)
 
-        with mock.patch("google.cloud.spanner_v1.transaction.Transaction.commit"):
-            connection.rollback()
+        connection.rollback()
 
         self.assertEqual(len(connection._statements), 0)
 
-    def test_retry_transaction(self):
+    def test_retry_transaction_w_checksum_match(self):
         """Check retrying an aborted transaction."""
         from google.cloud.spanner_dbapi.checksum import ResultsChecksum
         from google.cloud.spanner_dbapi.cursor import Statement
 
         row = ["field1", "field2"]
         connection = self._make_connection()
-
         checksum = ResultsChecksum()
         checksum.consume_result(row)
+
         retried_checkum = ResultsChecksum()
+        run_mock = connection.run_statement = mock.Mock()
+        run_mock.return_value = ([row], retried_checkum)
 
         statement = Statement("SELECT 1", [], {}, checksum, False)
         connection._statements.append(statement)
 
         with mock.patch(
-            "google.cloud.spanner_dbapi.connection.Connection.run_statement",
-            return_value=([row], retried_checkum),
-        ) as run_mock:
-            with mock.patch(
-                "google.cloud.spanner_dbapi.connection._compare_checksums"
-            ) as compare_mock:
-                connection.retry_transaction()
+            "google.cloud.spanner_dbapi.connection._compare_checksums"
+        ) as compare_mock:
+            connection.retry_transaction()
 
-                compare_mock.assert_called_with(checksum, retried_checkum)
+        compare_mock.assert_called_with(checksum, retried_checkum)
+        run_mock.assert_called_with(statement, retried=True)
 
-            run_mock.assert_called_with(statement, retried=True)
-
-    def test_retry_transaction_checksum_mismatch(self):
+    def test_retry_transaction_w_checksum_mismatch(self):
         """
         Check retrying an aborted transaction
         with results checksums mismatch.
@@ -497,18 +428,17 @@ class TestConnection(unittest.TestCase):
         checksum = ResultsChecksum()
         checksum.consume_result(row)
         retried_checkum = ResultsChecksum()
+        run_mock = connection.run_statement = mock.Mock()
+        run_mock.return_value = ([retried_row], retried_checkum)
 
         statement = Statement("SELECT 1", [], {}, checksum, False)
         connection._statements.append(statement)
 
-        with mock.patch(
-            "google.cloud.spanner_dbapi.connection.Connection.run_statement",
-            return_value=([retried_row], retried_checkum),
-        ):
-            with self.assertRaises(RetryAborted):
-                connection.retry_transaction()
+        with self.assertRaises(RetryAborted):
+            connection.retry_transaction()
 
-    def test_commit_retry_aborted_statements(self):
+    @mock.patch("google.cloud.spanner_v1.Client")
+    def test_commit_retry_aborted_statements(self, mock_client):
         """Check that retried transaction executing the same statements."""
         from google.api_core.exceptions import Aborted
         from google.cloud.spanner_dbapi.checksum import ResultsChecksum
@@ -516,13 +446,8 @@ class TestConnection(unittest.TestCase):
         from google.cloud.spanner_dbapi.cursor import Statement
 
         row = ["field1", "field2"]
-        with mock.patch(
-            "google.cloud.spanner_v1.instance.Instance.exists", return_value=True,
-        ):
-            with mock.patch(
-                "google.cloud.spanner_v1.database.Database.exists", return_value=True,
-            ):
-                connection = connect("test-instance", "test-database")
+
+        connection = connect("test-instance", "test-database")
 
         cursor = connection.cursor()
         cursor._checksum = ResultsChecksum()
@@ -530,19 +455,15 @@ class TestConnection(unittest.TestCase):
 
         statement = Statement("SELECT 1", [], {}, cursor._checksum, False)
         connection._statements.append(statement)
-        connection._transaction = mock.Mock(rolled_back=False, committed=False)
+        mock_transaction = mock.Mock(rolled_back=False, committed=False)
+        connection._transaction = mock_transaction
+        mock_transaction.commit.side_effect = [Aborted("Aborted"), None]
+        run_mock = connection.run_statement = mock.Mock()
+        run_mock.return_value = ([row], ResultsChecksum())
 
-        with mock.patch.object(
-            connection._transaction, "commit", side_effect=(Aborted("Aborted"), None),
-        ):
-            with mock.patch(
-                "google.cloud.spanner_dbapi.connection.Connection.run_statement",
-                return_value=([row], ResultsChecksum()),
-            ) as run_mock:
+        connection.commit()
 
-                connection.commit()
-
-                run_mock.assert_called_with(statement, retried=True)
+        run_mock.assert_called_with(statement, retried=True)
 
     def test_retry_transaction_drop_transaction(self):
         """
@@ -558,7 +479,8 @@ class TestConnection(unittest.TestCase):
         connection.retry_transaction()
         self.assertIsNone(connection._transaction)
 
-    def test_retry_aborted_retry(self):
+    @mock.patch("google.cloud.spanner_v1.Client")
+    def test_retry_aborted_retry(self, mock_client):
         """
         Check that in case of a retried transaction failed,
         the connection will retry it once again.
@@ -570,13 +492,7 @@ class TestConnection(unittest.TestCase):
 
         row = ["field1", "field2"]
 
-        with mock.patch(
-            "google.cloud.spanner_v1.instance.Instance.exists", return_value=True,
-        ):
-            with mock.patch(
-                "google.cloud.spanner_v1.database.Database.exists", return_value=True,
-            ):
-                connection = connect("test-instance", "test-database")
+        connection = connect("test-instance", "test-database")
 
         cursor = connection.cursor()
         cursor._checksum = ResultsChecksum()
@@ -584,27 +500,19 @@ class TestConnection(unittest.TestCase):
 
         statement = Statement("SELECT 1", [], {}, cursor._checksum, False)
         connection._statements.append(statement)
-
         metadata_mock = mock.Mock()
         metadata_mock.trailing_metadata.return_value = {}
+        run_mock = connection.run_statement = mock.Mock()
+        run_mock.side_effect = [
+            Aborted("Aborted", errors=[metadata_mock]),
+            ([row], ResultsChecksum()),
+        ]
 
-        with mock.patch.object(
-            connection,
-            "run_statement",
-            side_effect=(
-                Aborted("Aborted", errors=[metadata_mock]),
-                ([row], ResultsChecksum()),
-            ),
-        ) as retry_mock:
+        connection.retry_transaction()
 
-            connection.retry_transaction()
-
-            retry_mock.assert_has_calls(
-                (
-                    mock.call(statement, retried=True),
-                    mock.call(statement, retried=True),
-                )
-            )
+        run_mock.assert_has_calls(
+            (mock.call(statement, retried=True), mock.call(statement, retried=True),)
+        )
 
     def test_retry_transaction_raise_max_internal_retries(self):
         """Check retrying raise an error of max internal retries."""
@@ -627,7 +535,8 @@ class TestConnection(unittest.TestCase):
 
         conn.MAX_INTERNAL_RETRIES = 50
 
-    def test_retry_aborted_retry_without_delay(self):
+    @mock.patch("google.cloud.spanner_v1.Client")
+    def test_retry_aborted_retry_without_delay(self, mock_client):
         """
         Check that in case of a retried transaction failed,
         the connection will retry it once again.
@@ -639,13 +548,7 @@ class TestConnection(unittest.TestCase):
 
         row = ["field1", "field2"]
 
-        with mock.patch(
-            "google.cloud.spanner_v1.instance.Instance.exists", return_value=True,
-        ):
-            with mock.patch(
-                "google.cloud.spanner_v1.database.Database.exists", return_value=True,
-            ):
-                connection = connect("test-instance", "test-database")
+        connection = connect("test-instance", "test-database")
 
         cursor = connection.cursor()
         cursor._checksum = ResultsChecksum()
@@ -653,29 +556,20 @@ class TestConnection(unittest.TestCase):
 
         statement = Statement("SELECT 1", [], {}, cursor._checksum, False)
         connection._statements.append(statement)
-
         metadata_mock = mock.Mock()
         metadata_mock.trailing_metadata.return_value = {}
+        run_mock = connection.run_statement = mock.Mock()
+        run_mock.side_effect = [
+            Aborted("Aborted", errors=[metadata_mock]),
+            ([row], ResultsChecksum()),
+        ]
+        connection._get_retry_delay = mock.Mock(return_value=False)
 
-        with mock.patch(
-            "google.cloud.spanner_dbapi.connection.Connection.run_statement",
-            side_effect=(
-                Aborted("Aborted", errors=[metadata_mock]),
-                ([row], ResultsChecksum()),
-            ),
-        ) as retry_mock:
-            with mock.patch(
-                "google.cloud.spanner_dbapi.connection._get_retry_delay",
-                return_value=False,
-            ):
-                connection.retry_transaction()
+        connection.retry_transaction()
 
-            retry_mock.assert_has_calls(
-                (
-                    mock.call(statement, retried=True),
-                    mock.call(statement, retried=True),
-                )
-            )
+        run_mock.assert_has_calls(
+            (mock.call(statement, retried=True), mock.call(statement, retried=True),)
+        )
 
     def test_retry_transaction_w_multiple_statement(self):
         """Check retrying an aborted transaction."""
@@ -693,19 +587,17 @@ class TestConnection(unittest.TestCase):
         statement1 = Statement("SELECT 2", [], {}, checksum, False)
         connection._statements.append(statement)
         connection._statements.append(statement1)
+        run_mock = connection.run_statement = mock.Mock()
+        run_mock.return_value = ([row], retried_checkum)
 
         with mock.patch(
-            "google.cloud.spanner_dbapi.connection.Connection.run_statement",
-            return_value=([row], retried_checkum),
-        ) as run_mock:
-            with mock.patch(
-                "google.cloud.spanner_dbapi.connection._compare_checksums"
-            ) as compare_mock:
-                connection.retry_transaction()
+            "google.cloud.spanner_dbapi.connection._compare_checksums"
+        ) as compare_mock:
+            connection.retry_transaction()
 
-                compare_mock.assert_called_with(checksum, retried_checkum)
+        compare_mock.assert_called_with(checksum, retried_checkum)
 
-            run_mock.assert_called_with(statement1, retried=True)
+        run_mock.assert_called_with(statement1, retried=True)
 
     def test_retry_transaction_w_empty_response(self):
         """Check retrying an aborted transaction."""
@@ -721,16 +613,14 @@ class TestConnection(unittest.TestCase):
 
         statement = Statement("SELECT 1", [], {}, checksum, False)
         connection._statements.append(statement)
+        run_mock = connection.run_statement = mock.Mock()
+        run_mock.return_value = ([row], retried_checkum)
 
         with mock.patch(
-            "google.cloud.spanner_dbapi.connection.Connection.run_statement",
-            return_value=(row, retried_checkum),
-        ) as run_mock:
-            with mock.patch(
-                "google.cloud.spanner_dbapi.connection._compare_checksums"
-            ) as compare_mock:
-                connection.retry_transaction()
+            "google.cloud.spanner_dbapi.connection._compare_checksums"
+        ) as compare_mock:
+            connection.retry_transaction()
 
-                compare_mock.assert_called_with(checksum, retried_checkum)
+        compare_mock.assert_called_with(checksum, retried_checkum)
 
-            run_mock.assert_called_with(statement, retried=True)
+        run_mock.assert_called_with(statement, retried=True)

--- a/tests/unit/spanner_dbapi/test_cursor.py
+++ b/tests/unit/spanner_dbapi/test_cursor.py
@@ -70,16 +70,11 @@ class TestCursor(unittest.TestCase):
         with self.assertRaises(InterfaceError):
             cursor.callproc(procname=None)
 
-    def test_close(self):
+    @mock.patch("google.cloud.spanner_v1.Client")
+    def test_close(self, mock_client):
         from google.cloud.spanner_dbapi import connect, InterfaceError
 
-        with mock.patch(
-            "google.cloud.spanner_v1.instance.Instance.exists", return_value=True
-        ):
-            with mock.patch(
-                "google.cloud.spanner_v1.database.Database.exists", return_value=True
-            ):
-                connection = connect(self.INSTANCE, self.DATABASE)
+        connection = connect(self.INSTANCE, self.DATABASE)
 
         cursor = connection.cursor()
         self.assertFalse(cursor.is_closed)
@@ -87,6 +82,7 @@ class TestCursor(unittest.TestCase):
         cursor.close()
 
         self.assertTrue(cursor.is_closed)
+
         with self.assertRaises(InterfaceError):
             cursor.execute("SELECT * FROM database")
 
@@ -276,17 +272,12 @@ class TestCursor(unittest.TestCase):
             with self.assertRaises(OperationalError):
                 cursor.execute(sql="sql")
 
-    def test_executemany_on_closed_cursor(self):
+    @mock.patch("google.cloud.spanner_v1.Client")
+    def test_executemany_on_closed_cursor(self, mock_client):
         from google.cloud.spanner_dbapi import InterfaceError
         from google.cloud.spanner_dbapi import connect
 
-        with mock.patch(
-            "google.cloud.spanner_v1.instance.Instance.exists", return_value=True
-        ):
-            with mock.patch(
-                "google.cloud.spanner_v1.database.Database.exists", return_value=True
-            ):
-                connection = connect("test-instance", "test-database")
+        connection = connect("test-instance", "test-database")
 
         cursor = connection.cursor()
         cursor.close()
@@ -294,35 +285,25 @@ class TestCursor(unittest.TestCase):
         with self.assertRaises(InterfaceError):
             cursor.executemany("""SELECT * FROM table1 WHERE "col1" = @a1""", ())
 
-    def test_executemany_DLL(self):
+    @mock.patch("google.cloud.spanner_v1.Client")
+    def test_executemany_DLL(self, mock_client):
         from google.cloud.spanner_dbapi import connect, ProgrammingError
 
-        with mock.patch(
-            "google.cloud.spanner_v1.instance.Instance.exists", return_value=True,
-        ):
-            with mock.patch(
-                "google.cloud.spanner_v1.database.Database.exists", return_value=True,
-            ):
-                connection = connect("test-instance", "test-database")
+        connection = connect("test-instance", "test-database")
 
         cursor = connection.cursor()
 
         with self.assertRaises(ProgrammingError):
             cursor.executemany("""DROP DATABASE database_name""", ())
 
-    def test_executemany(self):
+    @mock.patch("google.cloud.spanner_v1.Client")
+    def test_executemany(self, mock_client):
         from google.cloud.spanner_dbapi import connect
 
         operation = """SELECT * FROM table1 WHERE "col1" = @a1"""
         params_seq = ((1,), (2,))
 
-        with mock.patch(
-            "google.cloud.spanner_v1.instance.Instance.exists", return_value=True
-        ):
-            with mock.patch(
-                "google.cloud.spanner_v1.database.Database.exists", return_value=True
-            ):
-                connection = connect("test-instance", "test-database")
+        connection = connect("test-instance", "test-database")
 
         cursor = connection.cursor()
         cursor._result_set = [1, 2, 3]
@@ -561,7 +542,8 @@ class TestCursor(unittest.TestCase):
             )
             self.assertEqual(result, expected)
 
-    def test_peek_iterator_aborted(self):
+    @mock.patch("google.cloud.spanner_v1.Client")
+    def test_peek_iterator_aborted(self, mock_client):
         """
         Checking that an Aborted exception is retried in case it happened
         while streaming the first element with a PeekIterator.
@@ -569,13 +551,7 @@ class TestCursor(unittest.TestCase):
         from google.api_core.exceptions import Aborted
         from google.cloud.spanner_dbapi.connection import connect
 
-        with mock.patch(
-            "google.cloud.spanner_v1.instance.Instance.exists", return_value=True,
-        ):
-            with mock.patch(
-                "google.cloud.spanner_v1.database.Database.exists", return_value=True,
-            ):
-                connection = connect("test-instance", "test-database")
+        connection = connect("test-instance", "test-database")
 
         cursor = connection.cursor()
         with mock.patch(
@@ -593,7 +569,8 @@ class TestCursor(unittest.TestCase):
 
                 retry_mock.assert_called_with()
 
-    def test_peek_iterator_aborted_autocommit(self):
+    @mock.patch("google.cloud.spanner_v1.Client")
+    def test_peek_iterator_aborted_autocommit(self, mock_client):
         """
         Checking that an Aborted exception is retried in case it happened while
         streaming the first element with a PeekIterator in autocommit mode.
@@ -601,13 +578,7 @@ class TestCursor(unittest.TestCase):
         from google.api_core.exceptions import Aborted
         from google.cloud.spanner_dbapi.connection import connect
 
-        with mock.patch(
-            "google.cloud.spanner_v1.instance.Instance.exists", return_value=True,
-        ):
-            with mock.patch(
-                "google.cloud.spanner_v1.database.Database.exists", return_value=True,
-            ):
-                connection = connect("test-instance", "test-database")
+        connection = connect("test-instance", "test-database")
 
         connection.autocommit = True
         cursor = connection.cursor()
@@ -629,19 +600,14 @@ class TestCursor(unittest.TestCase):
 
                 retry_mock.assert_called_with()
 
-    def test_fetchone_retry_aborted(self):
+    @mock.patch("google.cloud.spanner_v1.Client")
+    def test_fetchone_retry_aborted(self, mock_client):
         """Check that aborted fetch re-executing transaction."""
         from google.api_core.exceptions import Aborted
         from google.cloud.spanner_dbapi.checksum import ResultsChecksum
         from google.cloud.spanner_dbapi.connection import connect
 
-        with mock.patch(
-            "google.cloud.spanner_v1.instance.Instance.exists", return_value=True,
-        ):
-            with mock.patch(
-                "google.cloud.spanner_v1.database.Database.exists", return_value=True,
-            ):
-                connection = connect("test-instance", "test-database")
+        connection = connect("test-instance", "test-database")
 
         cursor = connection.cursor()
         cursor._checksum = ResultsChecksum()
@@ -658,7 +624,8 @@ class TestCursor(unittest.TestCase):
 
                 retry_mock.assert_called_with()
 
-    def test_fetchone_retry_aborted_statements(self):
+    @mock.patch("google.cloud.spanner_v1.Client")
+    def test_fetchone_retry_aborted_statements(self, mock_client):
         """Check that retried transaction executing the same statements."""
         from google.api_core.exceptions import Aborted
         from google.cloud.spanner_dbapi.checksum import ResultsChecksum
@@ -666,13 +633,7 @@ class TestCursor(unittest.TestCase):
         from google.cloud.spanner_dbapi.cursor import Statement
 
         row = ["field1", "field2"]
-        with mock.patch(
-            "google.cloud.spanner_v1.instance.Instance.exists", return_value=True,
-        ):
-            with mock.patch(
-                "google.cloud.spanner_v1.database.Database.exists", return_value=True,
-            ):
-                connection = connect("test-instance", "test-database")
+        connection = connect("test-instance", "test-database")
 
         cursor = connection.cursor()
         cursor._checksum = ResultsChecksum()
@@ -694,7 +655,8 @@ class TestCursor(unittest.TestCase):
 
                 run_mock.assert_called_with(statement, retried=True)
 
-    def test_fetchone_retry_aborted_statements_checksums_mismatch(self):
+    @mock.patch("google.cloud.spanner_v1.Client")
+    def test_fetchone_retry_aborted_statements_checksums_mismatch(self, mock_client):
         """Check transaction retrying with underlying data being changed."""
         from google.api_core.exceptions import Aborted
         from google.cloud.spanner_dbapi.exceptions import RetryAborted
@@ -705,13 +667,7 @@ class TestCursor(unittest.TestCase):
         row = ["field1", "field2"]
         row2 = ["updated_field1", "field2"]
 
-        with mock.patch(
-            "google.cloud.spanner_v1.instance.Instance.exists", return_value=True,
-        ):
-            with mock.patch(
-                "google.cloud.spanner_v1.database.Database.exists", return_value=True,
-            ):
-                connection = connect("test-instance", "test-database")
+        connection = connect("test-instance", "test-database")
 
         cursor = connection.cursor()
         cursor._checksum = ResultsChecksum()
@@ -734,19 +690,14 @@ class TestCursor(unittest.TestCase):
 
                 run_mock.assert_called_with(statement, retried=True)
 
-    def test_fetchall_retry_aborted(self):
+    @mock.patch("google.cloud.spanner_v1.Client")
+    def test_fetchall_retry_aborted(self, mock_client):
         """Check that aborted fetch re-executing transaction."""
         from google.api_core.exceptions import Aborted
         from google.cloud.spanner_dbapi.checksum import ResultsChecksum
         from google.cloud.spanner_dbapi.connection import connect
 
-        with mock.patch(
-            "google.cloud.spanner_v1.instance.Instance.exists", return_value=True,
-        ):
-            with mock.patch(
-                "google.cloud.spanner_v1.database.Database.exists", return_value=True,
-            ):
-                connection = connect("test-instance", "test-database")
+        connection = connect("test-instance", "test-database")
 
         cursor = connection.cursor()
         cursor._checksum = ResultsChecksum()
@@ -763,7 +714,8 @@ class TestCursor(unittest.TestCase):
 
                 retry_mock.assert_called_with()
 
-    def test_fetchall_retry_aborted_statements(self):
+    @mock.patch("google.cloud.spanner_v1.Client")
+    def test_fetchall_retry_aborted_statements(self, mock_client):
         """Check that retried transaction executing the same statements."""
         from google.api_core.exceptions import Aborted
         from google.cloud.spanner_dbapi.checksum import ResultsChecksum
@@ -771,13 +723,7 @@ class TestCursor(unittest.TestCase):
         from google.cloud.spanner_dbapi.cursor import Statement
 
         row = ["field1", "field2"]
-        with mock.patch(
-            "google.cloud.spanner_v1.instance.Instance.exists", return_value=True,
-        ):
-            with mock.patch(
-                "google.cloud.spanner_v1.database.Database.exists", return_value=True,
-            ):
-                connection = connect("test-instance", "test-database")
+        connection = connect("test-instance", "test-database")
 
         cursor = connection.cursor()
         cursor._checksum = ResultsChecksum()
@@ -798,7 +744,8 @@ class TestCursor(unittest.TestCase):
 
                 run_mock.assert_called_with(statement, retried=True)
 
-    def test_fetchall_retry_aborted_statements_checksums_mismatch(self):
+    @mock.patch("google.cloud.spanner_v1.Client")
+    def test_fetchall_retry_aborted_statements_checksums_mismatch(self, mock_client):
         """Check transaction retrying with underlying data being changed."""
         from google.api_core.exceptions import Aborted
         from google.cloud.spanner_dbapi.exceptions import RetryAborted
@@ -809,13 +756,7 @@ class TestCursor(unittest.TestCase):
         row = ["field1", "field2"]
         row2 = ["updated_field1", "field2"]
 
-        with mock.patch(
-            "google.cloud.spanner_v1.instance.Instance.exists", return_value=True,
-        ):
-            with mock.patch(
-                "google.cloud.spanner_v1.database.Database.exists", return_value=True,
-            ):
-                connection = connect("test-instance", "test-database")
+        connection = connect("test-instance", "test-database")
 
         cursor = connection.cursor()
         cursor._checksum = ResultsChecksum()
@@ -838,19 +779,14 @@ class TestCursor(unittest.TestCase):
 
                 run_mock.assert_called_with(statement, retried=True)
 
-    def test_fetchmany_retry_aborted(self):
+    @mock.patch("google.cloud.spanner_v1.Client")
+    def test_fetchmany_retry_aborted(self, mock_client):
         """Check that aborted fetch re-executing transaction."""
         from google.api_core.exceptions import Aborted
         from google.cloud.spanner_dbapi.checksum import ResultsChecksum
         from google.cloud.spanner_dbapi.connection import connect
 
-        with mock.patch(
-            "google.cloud.spanner_v1.instance.Instance.exists", return_value=True,
-        ):
-            with mock.patch(
-                "google.cloud.spanner_v1.database.Database.exists", return_value=True,
-            ):
-                connection = connect("test-instance", "test-database")
+        connection = connect("test-instance", "test-database")
 
         cursor = connection.cursor()
         cursor._checksum = ResultsChecksum()
@@ -867,7 +803,8 @@ class TestCursor(unittest.TestCase):
 
                 retry_mock.assert_called_with()
 
-    def test_fetchmany_retry_aborted_statements(self):
+    @mock.patch("google.cloud.spanner_v1.Client")
+    def test_fetchmany_retry_aborted_statements(self, mock_client):
         """Check that retried transaction executing the same statements."""
         from google.api_core.exceptions import Aborted
         from google.cloud.spanner_dbapi.checksum import ResultsChecksum
@@ -875,13 +812,7 @@ class TestCursor(unittest.TestCase):
         from google.cloud.spanner_dbapi.cursor import Statement
 
         row = ["field1", "field2"]
-        with mock.patch(
-            "google.cloud.spanner_v1.instance.Instance.exists", return_value=True,
-        ):
-            with mock.patch(
-                "google.cloud.spanner_v1.database.Database.exists", return_value=True,
-            ):
-                connection = connect("test-instance", "test-database")
+        connection = connect("test-instance", "test-database")
 
         cursor = connection.cursor()
         cursor._checksum = ResultsChecksum()
@@ -903,7 +834,8 @@ class TestCursor(unittest.TestCase):
 
                 run_mock.assert_called_with(statement, retried=True)
 
-    def test_fetchmany_retry_aborted_statements_checksums_mismatch(self):
+    @mock.patch("google.cloud.spanner_v1.Client")
+    def test_fetchmany_retry_aborted_statements_checksums_mismatch(self, mock_client):
         """Check transaction retrying with underlying data being changed."""
         from google.api_core.exceptions import Aborted
         from google.cloud.spanner_dbapi.exceptions import RetryAborted
@@ -914,13 +846,7 @@ class TestCursor(unittest.TestCase):
         row = ["field1", "field2"]
         row2 = ["updated_field1", "field2"]
 
-        with mock.patch(
-            "google.cloud.spanner_v1.instance.Instance.exists", return_value=True,
-        ):
-            with mock.patch(
-                "google.cloud.spanner_v1.database.Database.exists", return_value=True,
-            ):
-                connection = connect("test-instance", "test-database")
+        connection = connect("test-instance", "test-database")
 
         cursor = connection.cursor()
         cursor._checksum = ResultsChecksum()
@@ -943,7 +869,8 @@ class TestCursor(unittest.TestCase):
 
                 run_mock.assert_called_with(statement, retried=True)
 
-    def test_ddls_with_semicolon(self):
+    @mock.patch("google.cloud.spanner_v1.Client")
+    def test_ddls_with_semicolon(self, mock_client):
         """
         Check that one script with several DDL statements separated
         with semicolons is splitted into several DDLs.
@@ -963,13 +890,7 @@ class TestCursor(unittest.TestCase):
             "DROP TABLE table_name",
         ]
 
-        with mock.patch(
-            "google.cloud.spanner_v1.instance.Instance.exists", return_value=True,
-        ):
-            with mock.patch(
-                "google.cloud.spanner_v1.database.Database.exists", return_value=True,
-            ):
-                connection = connect("test-instance", "test-database")
+        connection = connect("test-instance", "test-database")
 
         cursor = connection.cursor()
         cursor.execute(

